8302744: Refactor Hotspot container detection code

jerboaa opened the pull request:
Please review this container detection code refactoring in hotspot. The main point of this is to

- get rid of the `GET_CONTAINER_INFO` macros which hide too many things under the hood
- prevent refactoring of the code (since `GET_CONTAINER_INFO` macros short-return and are therefore not portable; at least not without some risk)
- make the code easier to understand
- allow for better testing via `gtest`
- separate multi-line parsing from single line parsing for clarity.

Testing:
- [x] GHA
- [x] `gtest:cgroupTest` tests
- [x] Container tests on Linux with cgroup v1 (legacy) and cgroup v2. All pass.

Thoughts?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302744](https://bugs.openjdk.org/browse/JDK-8302744): Refactor Hotspot container detection code (**Enhancement** - P4)


### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19060/head:pull/19060` \
`$ git checkout pull/19060`

Update a local copy of the PR: \
`$ git checkout pull/19060` \
`$ git pull https://git.openjdk.org/jdk.git pull/19060/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19060`

View PR using the GUI difftool: \
`$ git pr show -t 19060`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19060.diff">https://git.openjdk.org/jdk/pull/19060.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19060#issuecomment-2092616937)

jdksjolen commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -154,63 +183,44 @@ template <typename T> int subsystem_file_line_contents(CgroupController* c,
-}
+PRAGMA_DIAG_PUSH
+PRAGMA_FORMAT_NONLITERAL_IGNORED
+template <typename T> int cg_file_multi_line_ctrl(CgroupController* c,
```
This always bothered me: If the first argument is a pointer to a `CgroupController`, then why isn't this just a method for that object? No need to for the `c == nullptr` check then either.

jerboaa replied:
Good question. I hadn't considered it. Let me explore that option a bit and I'll get back to you.

jdksjolen reviewed (commented):
Hi @jerboaa ,

I'm happy to see effort being put towards  a cleanup of this code. I'll help out with reviewing next week.

jdksjolen reviewed (commented):
Hi Severin,

I've gone through a bit more of the code now.

I'm really wondering if simply doing the macro expansion is the right way to go here. So many of these do the same thing, isn't there a better pattern? I'm thinking something like this:

1. Change the calling convention to taking an out-parameter reference and return an int for error

```c++
int CgroupV1MemoryController::uses_mem_hierarchy(jlong& use_hierarchy) {
  jlong temp;
  int err = cg_file_contents_ctrl(this, "/memory.use_hierarchy", JLONG_FORMAT, &temp);
  if (err != 0) {
    log_trace(os, container)("Use Hierarchy is: %d", OSCONTAINER_ERROR);
    return err;
  }
  use_hierarchy = temp;
  log_trace(os, container)("Use Hierarchy is: " JLONG_FORMAT, use_hierarchy);
  return 0;
}
```

2. Abstract out the message such that it is obvious when it failed and not:

```c++
int CgroupV1MemoryController::uses_mem_hierarchy(jlong& use_hierarchy) {
  const char* str = "Use Hierarchy";
  jlong temp;
  int err = cg_file_contents_ctrl(this, "/memory.use_hierarchy", JLONG_FORMAT, &temp);
  if (err != 0) {
    log_trace(os, container)("%s failed: %d", str, OSCONTAINER_ERROR);
    return err;
  }
  log_trace(os, container)("%s is: " JLONG_FORMAT, str, use_hierarchy);
  return 0;
}
```

3. Port this pattern over to other similar methods and introduce a common macro for creating them. Here's some make believe code

```c++
#define V1_SIMPLE_ACCESSORS_LIST \
SIMPLE_ACCESSOR(julong, kernel_memory_usage_in_bytes,  _memory->controller(), "/memory.kmem.limit_in_bytes", "Kernel Memory Limit", JULONG_FORMAT, kmem_limit) \
SIMPLE_ACCESSOR(jlong, uses_mem_hierarchy, this,  "/memory.use_hierarchy", "Use Hierarchy", JLONG_FORMAT, use_hierarchy) \
SIMPLE_ACCESSOR(jlong, read_mem_swappiness, _memory->controller(), "/memory.swappiness", "Swappiness", JULONG_FORMAT, swappiness)

#define SIMPLE_ACCESSOR(ret_type, name, accessor, filename, logformat, logformat2, retname) \
int CgroupV1Subsystem::name(ret_type& retname) { \
  ret_type temp; \
int err = cg_file_contents_ctrl(accessor, filename, logformat2, &temp); \
if (err != 0) { log_debug(os)("%s: failed %d" logformat, err); return err; } \
 log_debug(os)("%s: failed " logformat2 logformat, temp); \
 retname = temp; \
 return 0; }
V1_SIMPLE_ACCESSORS_LIST
#undef SIMPLE_ACCESSOR
```

Which generates the following code (checked using `gcc -E` and formatted using `clang-format`)

```c++
int CgroupV1Subsystem::kernel_memory_usage_in_bytes(julong &kmem_limit) {
  julong temp;
  int err = cg_file_contents_ctrl(_memory->controller(),
                                  "/memory.kmem.limit_in_bytes", JULONG_FORMAT,
                                  &temp);
  if (err != 0) {
    log_debug(os)("%s: failed %d"
                  "Kernel Memory Limit",
                  err);
    return err;
  }
  log_debug(os)("%s: failed " JULONG_FORMAT "Kernel Memory Limit", temp);
  kmem_limit = temp;
  return 0;
}
int CgroupV1Subsystem::uses_mem_hierarchy(jlong &use_hierarchy) {
  jlong temp;
  int err =
      cg_file_contents_ctrl(this, "/memory.use_hierarchy", JLONG_FORMAT, &temp);
  if (err != 0) {
    log_debug(os)("%s: failed %d"
                  "Use Hierarchy",
                  err);
    return err;
  }
  log_debug(os)("%s: failed " JLONG_FORMAT "Use Hierarchy", temp);
  use_hierarchy = temp;
  return 0;
}
int CgroupV1Subsystem::read_mem_swappiness(jlong &swappiness) {
  jlong temp;
  int err = cg_file_contents_ctrl(_memory->controller(), "/memory.swappiness",
                                  JULONG_FORMAT, &temp);
  if (err != 0) {
    log_debug(os)("%s: failed %d"
                  "Swappiness",
                  err);
    return err;
  }
  log_debug(os)("%s: failed " JULONG_FORMAT "Swappiness", temp);
  swappiness = temp;
  return 0;
}
```

Some of these methods do some more complex logic, I've noticed, so perhaps there isn't a silver bullet. Anyway, let me know what you think of this, it'd be good to simplify this code.

tstuefe reviewed (commented):
Hi Severin,

One thing I always disliked is that we wrap the scanf in a wrapper. That does cost us safety. We don't pass in format string literals, therefore the compiler cannot check the format strings for us.

Your patch makes it slightly worse: the old version just happened to use the log_fmt string in UL log lines as part of the GET_CONTAINER_INFO macro. They expanded as literals at the call site. This caused format string errors at least for wrong log format strings. Since you now moved the log lines the templatized function (or, removed them?), we don't get these warnings anymore.

Apart from the format string problem:

We have a templatized function that abstracts away the return type of the scanf operation. Okay. But we still manually hand in the scanf format string. So, nothing is gained from the templatization (before your patch and now).

Looking closely, afaics we only have a few cases:
a) %1023s, return a string
b) %ul, return 64bit unsigned
c) %d, return signed int (only found cpu-weigth, and documentation did not indicate this could be negative)
d) cpu-max is the odd duck out, consisting of a string and a number (eg "max 100000"). We ignore the string, just return the number.

case (b) and (c) could be handled as one, since you can read an unsigned also with ulong format.

So, this means we only have two cases - reading a string (of 1K length) and an unsigned numeric. And then the weird cpu-max thing. So I wonder whether we really need a templatized scan function at all. Why not just:
- a function scanning a 64bit unsigned from a file (I think even have that already)
- a function scanning a string from a file
- and the cpu-max thing could be handled by reading the line as string, then parsing the string with sscanf.

That would cut down a lot of complexity. And we would get compiler warnings for the format strings, no need for the pragmas anymore. What do you think?

Cheers, Thomas

jerboaa commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -69,46 +69,81 @@
+    log_debug(os, container)("Open of file %s failed, %s", absolute_path, os::strerror(errno));
+    return OSCONTAINER_ERROR;
+  }
+
```
I was taking the advise of the `scanf` man page here:
```
       The  scanf()  family  of functions scans input like sscanf(3), but read
       from a FILE.  It is very difficult to use  these  functions  correctly,
       and  it  is preferable to read entire lines with fgets(3) or getline(3)
       and parse them later with sscanf(3) or more specialized functions  such
       as strtol(3).
```

jerboaa commented:
Thanks all for the reviews! All good suggestions, though I'm not sure I like the macro approach @jdksjolen suggested. It still comes at the cost of harder to read code. I'll do further exploration and will hopefully come back with a better approach.

jerboaa commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -69,46 +69,81 @@
-                                              const char *key,
-                                              const char *scan_fmt,
-                                              T returnval) {
+template <typename T> int __cg_file_contents_impl(const char *absolute_path,
+                                                  const char *scan_fmt,
```
Thanks! Will fix.

jerboaa commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -69,46 +69,81 @@
-                                              const char *key,
-                                              const char *scan_fmt,
-                                              T returnval) {
+template <typename T> int __cg_file_contents_impl(const char *absolute_path,
```
> What's the significance of the double underscore prefix in the naming

The idea is that it shouldn't be used directly, but `cg_file_contents_ctrl` be used instead. Maybe I can get rid of this ugliness in the next iteration.

jankratochvil replied:
https://stackoverflow.com/a/224420/2995591

jerboaa commented on src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:
```diff
@@ -75,8 +75,13 @@ void CgroupV1Controller::set_subsystem_path(char *cgroup_path) {
  */
 jlong CgroupV1MemoryController::uses_mem_hierarchy() {
-  GET_CONTAINER_INFO(jlong, this, "/memory.use_hierarchy",
-                    "Use Hierarchy is: ", JLONG_FORMAT, JLONG_FORMAT, use_hierarchy);
+  jlong use_hierarchy;
+  int err = cg_file_contents_ctrl(this, "/memory.use_hierarchy", JLONG_FORMAT, &use_hierarchy);
+  if (err != 0) {
+    log_trace(os, container)("Use Hierarchy is: %d", OSCONTAINER_ERROR);
+    return (jlong)OSCONTAINER_ERROR;
+  }
+  log_trace(os, container)("Use Hierarchy is: " JLONG_FORMAT, use_hierarchy);
```
The idea would be for this use_hierarchy parsing to go away after https://bugs.openjdk.org/browse/JDK-8322420

jdksjolen commented:
> Thanks all for the reviews! All good suggestions, though I'm not sure I like the macro approach @jdksjolen suggested. It still comes at the cost of harder to read code. I'll do further exploration and will hopefully come back with a better approach.

Yeah, the X-macro approach may not be the best either. This is clearly a trade off between ease of reading one function, and the ease of making many extremely similar functions with few lines of code.

jerboaa commented:
> > Thanks all for the reviews! All good suggestions, though I'm not sure I like the macro approach @jdksjolen suggested. It still comes at the cost of harder to read code. I'll do further exploration and will hopefully come back with a better approach.
> 
> Yeah, the X-macro approach may not be the best either. This is clearly a trade off between ease of reading one function, and the ease of making many extremely similar functions with few lines of code.

Yes. Ideally, I'd like to reduce the duplication some other way.

jerboaa commented:
@jdksjolen @tstuefe Please take another look. I've introduced two simple macros for "checked" reading and the primary reading functions are reading a string or an unsigned long. Please let me know what you think.

jerboaa commented on test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:
```diff
@@ -200,8 +200,8 @@ private static void testMemorySwapLimitSanity() throws Exception {
             // Either for cgroup v2: a_2) 0, or b_2) -2 on systems with swapaccount=0
-            .shouldMatch("Memory and Swap Limit is:.*(" + expectedTraceValue + "|-2|0)")
-            .shouldNotMatch("Memory and Swap Limit is:.*" + neg2InUnsignedLong);
+            .shouldMatch("(Memory and )?Swap Limit is:.*(" + expectedTraceValue + "|-2|0)")
+            .shouldNotMatch("(Memory and )?Swap Limit is:.*" + neg2InUnsignedLong);
```
This is a side-effect of reading the swap limit on cg v2 which only prints `Swap Limit is:` (fixing a FIXME along the way in the code), since it's only reading the swap value in that case. Note that for cg v2 the interface files for swap limit and memory limit are different files and for cg v2, only the swap limit is there (contrary to the cg v1 interface file which contains the memory and swap limit). Therefore I've done an optional match in the test.

jdksjolen commented on src/hotspot/os/linux/cgroupSubsystem_linux.cpp:
```diff
@@ -556,6 +556,158 @@ jlong CgroupSubsystem::memory_limit_in_bytes() {
+PRAGMA_DIAG_PUSH
+PRAGMA_FORMAT_NONLITERAL_IGNORED // Only string/number literal formats used. See read_*() functions.
+template <typename T>
+bool CgroupController::read_from_file(const char* filename, const char* scan_fmt, T result) {
```
`read_value_from_ file`?

jerboaa replied:
No longer there.

jdksjolen commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -69,148 +69,49 @@
-                                                                          \
-  log_trace(os, container)(logstring, variable);                          \
-}
+enum TupleValue { FIRST, SECOND };
```
Suggestion/nit: Enum class + `First`, `Second`. Your choice to keep it as is, however.

jdksjolen replied:
I really want a more descriptive name and/or comment for this.

jdksjolen reviewed (commented):
This looks pretty good. Just the right amount of macrology, to me. I'll do another sweep, including the tests, later on.

Thank you for working on this.

jankratochvil commented on src/hotspot/os/linux/cgroupSubsystem_linux.cpp:
```diff
@@ -556,6 +556,158 @@ jlong CgroupSubsystem::memory_limit_in_bytes() {
+  int matched = sscanf(line, scan_fmt, result);
+  if (matched == 1) {
+    return true;
+  } else {
+    log_debug(os, container)("Type %s not found in file %s", scan_fmt, absolute_path);
+  }
```
```suggestion
  }
  log_debug(os, container)("Type %s not found in file %s", scan_fmt, absolute_path);
```

tstuefe reviewed (commented):
@jerboaa 

A lot better! 

If you move the sscanf up to `read_(number|numerical_tuple_value)()`, things become even simpler and you can remove the templatization altogether.

Also, at the moment, when you call read_string, three different buffers are involved, and three buffer copies. And one of the buffers is malloc'ed, which is not ideal for a high frequency call. And it forces you to deal with freeing that buffer.

You have:

- `read_number` -> `read_from_file`
- `read_numerical_tuple_value` -> `read_from_file`
- `read_string` -> `read_from_file`

You can move the sscanf up to the callers:

- Let `read_from_file()` only read the first line from a file and return that. No sscanf. Let it take an output buffer (e.g. `char output[1024]`) as argument.
- Let `read_number()` call `read_from_file` with a local buffer, then call sscanf with its `JULONG_FORMAT`
- Let `read_numerical_tuple_value()` call `read_from_file` with a local buffer, then call sscanf with the TupleFormat
- Let `read_string` call `read_from_file` with a local buffer. No need for sscanf. It can just return the strduped buffer.

No need for templatization of `read_from_file` then.

An extension of that idea would be to not use strdup in `read_string` at all, but to require a caller-provided buffer. Then, we don't have to copy data around, and you could just remove `read_string` altogether, since it adds nothing of value. Callers can call `read_from_file` directly.

Removing the strdup from the read_string path would mean you can also remove all the `os::free` calls in limit_from_str. No need to think about who owns which buffers.

You would introduce local buffers in those functions that call read_string. In those that work on caller-provided buffers, return either the caller-provided buffer or NULL to indicate an error (which is the typical C-style pattern), e.g.:


```
char* CgroupV2Subsystem::pids_max_val(char buffer[1024]) {
  CONTAINER_READ_STRING_CHECKED(_unified, "/pids.max", "Maximum number of tasks", buffer);
  return buffer;
}

jlong CgroupV2Subsystem::pids_max() {
  char buffer[1024]
  char * pidsmax_str = pids_max_val(buffer);
  return limit_from_str(buffer);
}
```

Yet another way to avoid malloc would be using ResourceAreas. But that introduces dependencies to initialization, and I hesitate having cgroups handling depend on VM initialization.

jerboaa commented:
@tstuefe @jdksjolen Take 3 :) No more templates involved. PTAL when you get a chance. Thanks again for the feedback! Much appreciated.

tstuefe commented:
@jerboaa nice :) Will take a look next week

tstuefe commented on src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp:
```diff
@@ -294,15 +318,24 @@ char * CgroupV1Subsystem::cpu_cpuset_memory_nodes() {
+  julong quota;
+  bool is_ok = _cpu->controller()->
+                  read_number("/cpu.cfs_quota_us", &quota);
+  if (!is_ok) {
```
could probably scrap is_ok and just do the call in the condition

jdksjolen commented on src/hotspot/os/linux/cgroupSubsystem_linux.cpp:
```diff
@@ -556,21 +556,186 @@ jlong CgroupSubsystem::memory_limit_in_bytes() {
+  }
+  char token[1024];
+  int matched = -1;
+  switch(tup) {
```
Style: Add a space between switch and `(tup)`.

jdksjolen commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -69,148 +69,99 @@
+ * is <quota> and the second tuple value is <period>. We use this to map the
+ * tuple value to a constant string format for sscanf reading.
+ */
+enum class TupleValue { FIRST, SECOND };
```
Style, nit: No reason for all-caps for these names, can just be "First", "Second".

jdksjolen commented on src/hotspot/os/linux/cgroupV2Subsystem_linux.cpp:
```diff
@@ -140,21 +146,25 @@ jlong CgroupV2Subsystem::memory_max_usage_in_bytes() {
-                          "anon", JULONG_FORMAT, JULONG_FORMAT, rss);
-  return rss;
+  julong rss;
+  bool is_ok = _memory->controller()->
+                    read_numerical_key_value("/memory.stat", "anon", &rss);
+  if (!is_ok) {
```
Style, nit: A bit unconventional (for Hotspot) to line break after `->`, maybe this is preferable:

```c++
bool is_ok =
  _memory->controller()->read_numerical_key_value("/memory.stat", "anon", &rss);
```

jdksjolen replied:
Or each argument can be line-separated, just a couple of ideas.

jdksjolen reviewed (commented):
Just a couple of style comments from me. I'll give it another round of looking over soon.

jerboaa commented on src/hotspot/os/linux/cgroupSubsystem_linux.cpp:
```diff
@@ -556,21 +556,186 @@ jlong CgroupSubsystem::memory_limit_in_bytes() {
+    log_debug(os, container)("Empty file %s", absolute_path);
+    fclose(fp);
+    return false;
+  }
```
You are right. The loop conditional is sufficient.

jerboaa commented on src/hotspot/os/linux/cgroupSubsystem_linux.hpp:
```diff
@@ -69,148 +69,99 @@
+     * in buffer will contain the first 1023 characters of the string or
+     * up to the first new line character ('\n') whichever comes first.
+     */
+    bool read_string(const char* filename, char* buf);
```
OK. Went with the former (specifying the buffer size as well as the buffer).

jerboaa commented:
> This is a really nice cleanup. Some more remarks inline, but its nearing the finish line from my point of view.

Thanks for the review!
 
> Is the file move of test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp necessary? Difficult to see what changed.

It's not necessary, but tests for the same file have been split between two files which isn't very convenient. What changed is that I've taken existing tests and added some and adjusted to the new interface file parsing functions. It just seems appropriate since we are refactoring the product code anyway. Test coverage is the same (or better).

tstuefe reviewed (commented):
Final look at the gtests, small nits

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+  TestController(char *p) {
+    _path = p;
+  }
+  char* subsystem_path() override {
```
I'm not sure this is possible. The subsystem path is not set by the constructor. I'll leave that for another day.

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+  is_ok = controller->read_numerical_key_value(base_with_slash, "foo", &x);
+  EXPECT_FALSE(is_ok) << "key must be exact match";
+  EXPECT_EQ((julong)0xBAD, x) << "x must be unchanged";
+
```
OK.

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+  delete_file(test_file);
+}
+
+TEST(cgroupTest, read_number_null) {
```
The subsystem path is being set dynamically (possibly null), so the null check should stay.

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+  EXPECT_TRUE(ok) << "String parsing should have been successful";
+  EXPECT_STREQ("foo-bar", result) << "Expected strings to be equal";
+
+  result[0] = '\0';
+  fill_file(test_file, "1234");
+  ok = controller->read_string(base_with_slash, result, 1024);
+  EXPECT_TRUE(ok) << "String parsing should have been successful";
+  EXPECT_STREQ("1234", result) << "Expected strings to be equal";
+
+  // values with a space
+  result[0] = '\0';
+  fill_file(test_file, "abc def");
+  ok = controller->read_string(base_with_slash, result, 1024);
+  EXPECT_TRUE(ok) << "String parsing should have been successful";
+  EXPECT_STREQ("abc def", result) << "Expected strings to be equal";
```
Those two cases where there earlier and are valid positive checks (even if trivial). They returned `abc` for the space case prior this refactoring. Keeping the test shows this has been considered.

I've added the space with new-line case.

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+    too_large[i] = 'A' + (i % 26);
+  }
+  result[0] = '\0';
+  fill_file(test_file, too_large);
```
Fixed.

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+  EXPECT_TRUE(1023 == strlen(result)) << "Expected only the first 1023 chars to be read in";
+  for (int i = 0; i < 1023; i++) {
+    EXPECT_EQ(too_large[i], result[i]) << "Expected item at idx " << i << " to match";
+  }
```
Used `strncmp()` here.

jerboaa commented on test/hotspot/gtest/runtime/test_cgroupSubsystem_linux.cpp:
```diff
@@ -0,0 +1,454 @@
+
+class TestController : public CgroupController {
+private:
+  char* _path;
```
Used initializer list.

tstuefe reviewed (approved):
Looks okay to me now. If tests are green, good to go. Great cleanup.

jdksjolen reviewed (approved):
I went through the changes, and I also believe that this is mergeable. Thank you for your hard work!

jerboaa commented:
GHA test failure on Linux 32 bit x86 seems unrelated to this patch:
[serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage](https://github.com/jerboaa/jdk/actions/runs/9272985669#user-content-serviceability_jvmti_objectmonitorusage_objectmonitorusage)

jerboaa commented:
/integrate
